src/track/beatutils.cpp

```diff
@@ -5,44 +5,350 @@
  *      Author: vittorio
  */
 
-#include <algorithm>
-#include <QtDebug>
-#include <QString>
+#include "track/beatutils.h"
+
 #include <QList>
 #include <QMap>
+#include <QString>
+#include <QtDebug>
+#include <algorithm>
 
-#include "track/beatutils.h"
 #include "util/math.h"
+#include "util/descriptivestatistics.h"
+#include "util/windowedstatistics.h"
+
+namespace {
 
 // we are generous and assume the global_BPM to be at most 0.05 BPM far away
 // from the correct one
-#define BPM_ERROR 0.05
+constexpr double kMaxBpmError = 0.05;
+// When ironing the grid for long sequences of const tempo we use
+// a 25ms tolerence since this will sound unnoticed
+constexpr double kMaxSecsPhaseError = 0.025;
 
 // the raw beatgrid is divided into blocks of size N from which the local bpm is
 // computed. Tweaked from 8 to 12 which improves the BPM accuracy for 'problem songs'.
-#define N 12
+constexpr int kBeatsToCountTempo = 12;
 
 static bool sDebug = false;
 
 const double kCorrectBeatLocalBpmEpsilon = 0.05; //0.2;
 const int kHistogramDecimalPlaces = 2;
 const double kHistogramDecimalScale = pow(10.0, kHistogramDecimalPlaces);
 const double kBpmFilterTolerance = 1.0;
+// when comparing with two tempos should be the same we use this tolerence
+constexpr double kMaxDiffSameBpm = 0.3;
+
+QVector<double> makeQVector(QVector<double>::iterator begin, QVector<double>::iterator end) {
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+    return QVector<double>(begin, end);
+#else
+    return QVector<double>::fromStdVector(std::vector<double>(begin, end));
+#endif
+}
+
+} // namespace
+
+
+QMap<int, double> BeatUtils::findStableTempoRegions(
+        QMap<double, int> frequencyOfTempos, QList<double> tempoList) {
+    auto sortedTempoList = tempoList;
+    std::sort(sortedTempoList.begin(), sortedTempoList.end());
+    // We have to make sure we have odd numbers
+    if (!(sortedTempoList.size() % 2) and sortedTempoList.size() > 1) {
+        sortedTempoList.pop_back();
+    }
+    // Since we use the median as a guess first and last tempo
+    // And it can not have values outside from tempoFrequency
+    const double medianBpm = computeSampleMedian(sortedTempoList);
+    // Forming a meter perception takes a few seconds, so we assume sections of consistent
+    // metrical structure to be at least around 10s long. So we use a window of the double
+    // of that in our filtering.
+    int numberOfBeatsInFilteringWindow = (10 / (60 / medianBpm)) * 2;
+    if (!(numberOfBeatsInFilteringWindow % 2)) {
+        numberOfBeatsInFilteringWindow += 1;
+    }
+    auto tempoMedianFilter = MovingMedian(numberOfBeatsInFilteringWindow);
```

Comment:
```suggestion
    auto tempoMedianFilter = movingMedian(numberOfBeatsInFilteringWindow);
```
Use camelCase for function names.

Reply:
These are classes constructors, aren't those suppose to start with a Capital letter?

Reply:
sorry, I misunderstood then

Reply:
This doesn't build now. Please use the syntax
```
MovingMedian tempoMedianFilter(numberOfBeatsInFilteringWindow);
```
This makes it clear that MovingMedian is a class name even when looking at the diff on GitHub without syntax highlighting.

Reply:
@daschuer asked to write like this

Reply:
I don't care much either way, but @daschuer could you explain why you prefer it this way?

Reply:
With =  it is obvious that we assign a new variable. In the other version the variable name can be misread as a function call. We use this pattern in many places in Mixxx and I prefer this whenever possible. 

We had also a discussion about his, but I cannot find it. It looks like we failed to put that into the Wiki.